### PR TITLE
fix trim and trimkey : only leading and trailing characters are trimed

### PR DESCRIPTION
**Pull request by fbaligand**
Fix issue #10

### CODE REVIEW

**Comment by fbaligand** [comment]
@jordansissel @suyograo @purbon @ph 
Please, can you review this PR ?

**Comment by suyograo** [comment]
hmm is this considered a breaking change? The existing configs should work..

**Comment by fbaligand** [comment]
@suyograo 
There are 2 ways to see the thing :
- bugfix : trim means to remove only leading and trailing characters, not all characters.
- breaking change : existing configurations that use "trim" and "trim_keys" options to remove all characters (not only leading and trailing characters) won't have the same behaviour.

For my point of view, this is more a bugfix. Because the main use of these options is to remove blank or bracket chars that are around keys and values.
And you, what you think ?

**Comment by fbaligand** [comment]
@suyograo 
Tell me which kv version you want for this issue, and I will commit it.
Several people have put a +1 on issue #10, so I really believe it will be useful for logstash users to merge it and version a new release with that fix.

**Comment by jsvd** [comment]
@fbaligand what if we add a flag that keeps current behaviour?
My only concern is that, as is, we lose the ability to strip characters from keys and values. If we were to change trim, I'd suggest adding two more options called "strip" and "stripkey", and bumping the major version. what do you think?

**Comment by fbaligand** [comment]
Hi @jsvd ,

I think it is a good idea to have two different options to solve both needs !

So, because mutate plugin has already a strip option that removes only leading and trailing characters, and because ruby standard String.strip() method removes also only leading and trailing characters and finally to avoid a breaking change on kv plugin ; I suggest to keep trim and trimkey options as they are and to add 2 new "strip" and "stripkey" options that remove only leading and trailing characters.

What do you think about ?

**Comment by jsvd** [comment]
ughh english is tough. strip and trim are essentially synonyms in this situation.
We should probably come up with another name for the existing feature of removing all occurrences of characters from a string (e.g. eject, delete_char, remove_char, clean?), then rename the `trim` and `trimkey` to do what this PR implements

I'm ok with making `trim` do the right thing: trimming. I guess it's only debatable if the existing behaviour if a bug (so next version is a minor) or if this PR is a breaking change (next version is a major).
I'm tempted to just call this a bug fix + new feature, otherwise the current major version will never be fixed.

**Comment by fbaligand** [comment]
You're right : trim and strip are synonyms, and finally, it would be confusing to have these both option names on kv plugin.

So, to keep existing feature (delete all chars), I vote for "remove_char". Is it ok for you ?

Then, concerning "trim" and "trimkey", I'm not sure to understand : do you want to keep these names or do you want to rename to "strip" and "stripkey" ?

Finally, in all cases, I agree to make a new major release with both features.

**Comment by jsvd** [comment]
I think trim conveys better the message, so I'm ok with keeping trim and trimkey
on adding the new feature, maybe `remove_char_value` and `remove_char_key`? this is similar naming scheme to the existing `transform_key` and `transform_value`

+1 on major version

**Comment by fbaligand** [comment]
I agree with all that you said !

=> This will really be a major release :)

I will do the changes soon and update this PR.

**Comment by fbaligand** [comment]
Hi @jsvd,

I have committed all the changes !
Could you review it ?

**Comment by fbaligand** [comment]
Yes !
Happy to see it's merged !

So can you release 4.0.0 version and publish it to rubygems.org ?

**Comment by jsvd** [comment]
@fbaligand yep, just waiting for ci to complete, thank you for your stellar PR <3

**Comment by jsvd** [comment]
published https://rubygems.org/gems/logstash-filter-kv/versions/4.0.0

**Comment by fbaligand** [comment]
Great news !!!

**Comment by fbaligand** [comment]
@jsvd 
Could you close issue #10 ?

**Comment by fbaligand** [comment]
Thanks @jsvd .
One last thing : could you update official Logstash documentation on "master" branch ?

File: lib/logstash/filters/kv.rb [hunk 1]
```diff
@@ -41,29 +41,65 @@ class LogStash::Filters::KV < LogStash::Filters::Base
...
+  #
+  # For example, to trim `<`, `>`, `[`, `]` and `,` characters from values:
   # [source,ruby]
   #     filter {
   #       kv {
-  #         trim => "<>\[\],"
+  #         trim_value => "<>\[\],"
   #       }
   #     }
-  config :trim, :validate => :string
```
Comment (jordansissel) [inline lib/logstash/filters/kv.rb]: Hindsight:

THis should have been kept as a setting, but adding `:deprecated => "This setting has been renamed to 'trim_value'."`

File: lib/logstash/filters/kv.rb [hunk 2]
```diff
@@ -41,29 +41,65 @@ class LogStash::Filters::KV < LogStash::Filters::Base
...
+  #
+  # For example, to remove `<`, `>`, `[`, `]` and `,` characters from values:
   # [source,ruby]
   #     filter {
   #       kv {
-  #         trimkey => "<>\[\],"
+  #         remove_char_value => "<>\[\],"
   #       }
   #     }
-  config :trimkey, :validate => :string
```
Comment (jordansissel) [inline lib/logstash/filters/kv.rb]: Hindsight: This should have been marked deprecated, adding: `:deprecated => "This setting is renamed to 'remove_char_value'"`

**Comment by fbaligand** [comment]
You're right...
That said, the new released version is a major version (4.0.0), because it's a breaking version. 

One option would be to release an intermediate 3.x version with same 4.0.0 feature but with old options flagged "deprecated". 
What do you think about ?

**Pull request merged**